File: TangHongyin/construct_binary_tree_from_preorder_and_inorder_traversal.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):

    def traverse(self, preorder, inorder, pre_start, pre_end, in_start, in_end):
        root_value = preorder[pre_start]
        root = TreeNode(root_value)
        # the last position of the left tree in inorder array + 1
        inorder_left_end = inorder.index(root_value, in_start, in_end)
        # the last position of the left tree in preorder
        preorder_left_end = inorder_left_end - in_start + pre_start + 1
        # if left child exists
        if root_value != inorder[in_start]:
            root.left = self.traverse(preorder, inorder, pre_start + 1, preorder_left_end, in_start, inorder_left_end)
        inorder_right_start = inorder_left_end + 1
        preorder_right_start = preorder_left_end
        # if right child exists
        if root_value != inorder[in_end - 1]:
            root.right = self.traverse(preorder, inorder, preorder_right_start, pre_end, inorder_right_start, in_end)
        return root

    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        if preorder is None or len(preorder) == 0:
            return None
        return self.traverse(preorder, inorder, 0, len(preorder), 0, len(inorder))

if __name__ == "__main__":
    solution = Solution()
    solution.buildTree([1,2,4,8,5,3,6,9,7,10],[8,4,2,5,1,9,6,3,7,10])

File: Zhengyingying/src/com/zyy/week6/ConstructBinaryTreefromPreorderandInorderTraversal.java
package com.zyy.week6;

public class ConstructBinaryTreefromPreorderandInorderTraversal {
	// 深度优先搜索DFS
	public TreeNode buildTree(int[] preorder, int[] inorder) {
		return buildChildTree(0, preorder.length - 1, inorder, preorder, 0);
	}

	// 超时
	public TreeNode buildChildTree(int left, int right, int[] inorder,
			int[] preorder, int preIndex) {
		// 出口判断很重要
		if (preIndex >= preorder.length || left > right)
			return null;
		// 先根遍历中取出的是根元素
		TreeNode tree = new TreeNode(preorder[preIndex]);
		// 找以 tree为根的左右子树
		int index = 0;
		for (int i = left; i <= right; i++) {
			if (tree.val == inorder[i]) {
				index = i;
				break;
			}
		}
		// 左子树
		tree.left = buildChildTree(left, index - 1, inorder, preorder,
				preIndex + 1);
		// 右子树
		// 注意先序的位置
		tree.right = buildChildTree(index + 1, right, inorder, preorder,
				preIndex + index - left + 1);
		return tree;
	}

	public static void main(String[] args) {
		int[] inorder = new int[] { 1, 2, 3, 4 };
		int[] preorder = new int[] { 3, 1, 2, 4 };
		ConstructBinaryTreefromPreorderandInorderTraversal cbtpit = new ConstructBinaryTreefromPreorderandInorderTraversal();
		cbtpit.buildTree(preorder, inorder);

	}

}
